Support record_stream() for NJT

jbschlosser opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #137126
* #137125
* __->__ #137099

Does what it says on the tin. I believe the right behavior here is to ensure that `record_stream()` is called on all tensor components of the NJT to ensure they all live until stream computation is complete.

This is an ask from torchrec as the op is used there.

jbschlosser committed:
Support record_stream() for NJT

[ghstack-poisoned]

jbschlosser committed:
Update on "Support record_stream() for NJT"


Does what it says on the tin. I believe the right behavior here is to ensure that `record_stream()` is called on all tensor components of the NJT to ensure they all live until stream computation is complete.

This is an ask from torchrec as the op is used there.

[ghstack-poisoned]

jbschlosser committed:
Update on "Support record_stream() for NJT"


Does what it says on the tin. I believe the right behavior here is to ensure that `record_stream()` is called on all tensor components of the NJT to ensure they all live until stream computation is complete.

This is an ask from torchrec as the op is used there.

[ghstack-poisoned]

jbschlosser committed:
Update on "Support record_stream() for NJT"


Does what it says on the tin. I believe the right behavior here is to ensure that `record_stream()` is called on all tensor components of the NJT to ensure they all live until stream computation is complete.

This is an ask from torchrec as the op is used there.

[ghstack-poisoned]

jbschlosser committed:
Update on "Support record_stream() for NJT"


Does what it says on the tin. I believe the right behavior here is to ensure that `record_stream()` is called on all tensor components of the NJT to ensure they all live until stream computation is complete.

This is an ask from torchrec as the op is used there.

[ghstack-poisoned]

jbschlosser committed:
Update on "Support record_stream() for NJT"


Does what it says on the tin. I believe the right behavior here is to ensure that `record_stream()` is called on all tensor components of the NJT to ensure they all live until stream computation is complete.

This is an ask from torchrec as the op is used there.

[ghstack-poisoned]

ngimel commented on test/test_nestedtensor.py:
```diff
@@ -4231,6 +4231,26 @@ def test_threshold_backward(self, device):
+            nt.add_(1.0)
+            nt.record_stream(s)
+
+        self.assertEqual(nt, expected)
```
You don't wait on a side stream here, so the test passes by sheer accident. Also, this doesn't actually test `record_stream` functionality, which is that memory is not reused when the tensor goes out of scope, until the stream finishes. To test this I'd do something like
```
def fn():
    x = torch.ones(100, device="cuda") # make it nested
    data_ptr = x.data_ptr() # appropriate call for nested tensor
    s = torch.cuda.Stream()
    
    with torch.cuda.stream(s):
          # do something long, including sleep
          x.record_stream(s)
    return data_ptr

data_ptr = fn()
#allocate new tensor of same size, check that its data_ptr is not test tensor data_ptr

jbschlosser replied:
@ngimel I must be doing something wrong here; I can't get the test to fail without the `record_stream()` call. I'm using `torch._cuda.sleep()` to simulate doing something long and it appears to block?

jbschlosser committed:
Update on "Support record_stream() for NJT"


Does what it says on the tin. I believe the right behavior here is to ensure that `record_stream()` is called on all tensor components of the NJT to ensure they all live until stream computation is complete.

This is an ask from torchrec as the op is used there.

[ghstack-poisoned]

jbschlosser committed:
Update on "Support record_stream() for NJT"


Does what it says on the tin. I believe the right behavior here is to ensure that `record_stream()` is called on all tensor components of the NJT to ensure they all live until stream computation is complete.

This is an ask from torchrec as the op is used there.

[ghstack-poisoned]

jbschlosser commented:
@pytorchbot merge
